eclipse-theia/theia
# VS Code like theming (colors and icons)

#### What it does


fix #4831, fix #1486, fix #6300, fix #6747, fix #6776

- [x] (core) allow to register and use colors
  - Theia extensions can register new colors.
  - All colors are translated into css variables with `.` replaced by `-` and `--theia` prefix
  - Theia extensions can use registered colors via css or programatically from `ColorRegistrsy.getCurrentColor`
- [x] (plugin) hook [themes contribution point](https://code.visualstudio.com/api/references/contribution-points#contributes.themes) @akosyakov 
  - [x] json files
  - [x] tmTheme files
  - [x] Theia extensions should use `MonacoThemingService.register` to contribute new themes
- [x] (plugin) hook [colors contribution point](https://code.visualstudio.com/api/references/contribution-points#contributes.colors) @akosyakov 
- [x] (monaco) preserve themes between user sessions and reconnections @akosyakov 
- [x] register missing [VS Code colors](https://code.visualstudio.com/api/references/theme-color) from corresponding Theia extension, see https://github.com/eclipse-theia/theia/pull/6722
  - Base colors are reused from Monaco, one has to check what is missing by inspecting them.
  - restyle Theia extensions
  - remove existing css variables
- [x] define color theming coding guidelines
  - [x] make sure that colors are not defined in css anymore, i.e. `--theia-ui-font-color` and `--theia-layout-color` are not used
- [x] update CHANGELOG with breaking changes
- [x] (core) allow to register and select icon themes
  - [x] register existing theme
  - [x] register None theme to disable icon themes
  - [x] deprecate `name`, `description` and `icon` for tree nodes, use `LabelProvider` instead to avoid storing them in the local storage and compute synchronously only demand only for visible UI nodes
- [x] (plugin) hook [icon theme contribution point](https://code.visualstudio.com/api/extension-guides/icon-theme)
  - [x] icon definitions from paths
  - [x] preserve current icon theme between reloads
  - [x] icon definitions based on font definitions
  - [x] icon associations
  - [x] overriding icon associations for different color themes
- [x] (core) align theming commands and preferences with VS Code 
- [x] a CQ for copied code: https://dev.eclipse.org/ipzilla/show_bug.cgi?id=21223
- [x] review places where `FOLDER_ICON`  and `FILE_ICON` are used
- [x] local storage runs out of memory sometimes -> theme data are stored in indexeddb instead
- [x] debug console tab does not have icon
- [x] in offline mode for One Dark Pro theme status bar elements are bogusly highlighted while hovering over them
- [x] status bar coloring while debugging a program
- [x] custom default dark/light Theia themes, in light theme activity bar should be light as well
- [x] outline show duplicate elements

#### How to test


- uninstall `@theia/git` theia extension in the browser app
- install following VS Code extensions under `plugins` folder:
  - https://registry.npmjs.org/@theia/vscode-builtin-theme-defaults/-/vscode-builtin-theme-defaults-0.2.1.tgz to test default themes
  - https://registry.npmjs.org/@theia/vscode-builtin-vscode-theme-seti/-/vscode-builtin-vscode-theme-seti-0.3.0-next.2111655b67.tgz to test default VS Code icon theme
  - https://registry.npmjs.org/@theia/vscode-builtin-theme-red/-/vscode-builtin-theme-red-0.2.1.tgz and https://registry.npmjs.org/@theia/vscode-builtin-theme-tomorrow-night-blue/-/vscode-builtin-theme-tomorrow-night-blue-0.2.1.tgz to test custom themes (red and night)
  - https://registry.npmjs.org/@theia/vscode-builtin-git/-/vscode-builtin-git-0.2.1.tgz to test custom colors for git decorations in the navigator and scm views
  - https://github.com/microsoft/vscode-extension-samples/tree/master/theme-sample to test loading of color themes from tsTheme files
- check whether following UI respect icon theme switch:
  - [x] `@theia/editor`
    - [x] Quick Pick Editor Language
    - [x] Editor Quick Open
    - [x] Editor Title
      - [x] side areas
      - [x] main/bottom areas
    - [x] Diff Editor Title
  - [x] `@theia/filesystem`
    - [x] Quick File Open
    - [x] Open/Save File Dialog
  - [x] `@theia/navigator` - Navigator View
  - [x] `@theia/markers` - Problems View
  - [x] `@theia/git`
    - [x] SCM View
    - [x] Git Diff
    - [x] Git Commit Detail
    - [x] Git History
  - [x] `@theia/plugin-ext`
    - [x] SCM widget with git vscode built-in extension
    - [x] Tree Views
  - [x] `@theia/search-in-workspace` - Search View
  - [x] `@theia/workspace` - Quick Open Workspace
- https://marketplace.visualstudio.com/items?itemName=PKief.material-icon-theme to test hides `hidesExplorerArrows`

#### Review checklist


#### Reminder for reviewers

- as a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)
